### PR TITLE
Check the full vocab for grammar only if necessary

### PR DESCRIPTION
**Pull request by kalomaze**
Before, the GBNF Grammar sampler would parse the entire vocabulary before doing any sampling so that only candidates that were valid for the grammar are allowed. This is computationally expensive to run every single time, especially for complex grammars and for models with larger vocabulary sizes (llama 2 has 32,000 tokens, and recently Yi 34b has 64k!)

I have modified the sampling loop so that it does a 'first pass' for sampling. If Grammar is on, it will check from only the single token that it sampled from (`id`) to see if it matches the grammar rules by running the function. If it is set to -INF, that means we need to resample from scratch.

The function now includes a duplicate copy of the original logits for the 'resampling' process. If the chosen token violates the grammar rules, it will force the grammar rules and resample by calling the function recursively, this time with `is_resampling` set to True. In theory, this should greatly diminish the performance costs of grammar sampling by reducing the checks to only happen once for most tokens.

`main.cpp` and `infill.cpp` both had to be modified so that the function signatures being called were set to false, and idx was set to '0'. I'm not sure what the purpose of idx is, or what those files are both responsible for. Idx wasn't being called at all either in those instances, so I guess setting that to zero works?

### CODE REVIEW

**Comment by kalomaze** [comment]
It seems like I didn't break anything?

`main.exe -m "C:/Users/Kalomaze/Desktop/Local_LLM/Models/tinyllama-1.1b-intermediate-step-480k-1t.Q8_0.gguf" -n 256 --grammar-file "C:\Users\Kalomaze\Documents\GitHub\llama.cpp\grammars\json.gbnf" --prompt "An example of a file like this:"`

image

**Comment by kalomaze** [comment]
On Tinyllama, the sampling operation report goes from 15ms per token for 256 tokens, to 1ms per token for 256 tokens, when using the JSON grammar to test without and without the change.

File: common/sampling.h [hunk 1]
```diff
@@ -98,10 +98,11 @@ std::string llama_sampling_print(const llama_sampling_params & params);
...
 llama_token llama_sampling_sample(
-        struct llama_sampling_context * ctx_sampling,
-        struct llama_context * ctx_main,
-        struct llama_context * ctx_cfg,
-        int idx = 0);
+    struct llama_sampling_context * ctx_sampling,
+    struct llama_context * ctx_main,
+    struct llama_context * ctx_cfg,
+    const int idx,
+    bool is_resampling = false); // Add the new parameter with default value
```
[the remark this answers is not in the document]
Comment (ggerganov) [reply common/sampling.h]: Hm, not sure what you mean. The candidates array is part of the sampling context and it is cleared and populated on each call. The proposed code should work - does it not work?

File: common/sampling.cpp [hunk 2]
```diff
@@ -212,6 +223,29 @@ llama_token llama_sampling_sample(
...
+    if (ctx_sampling->grammar != NULL && !is_resampling) {
+        // Create an array with a single token data element for the sampled id
+        llama_token_data single_token_data = {id, logits[id], 0.0f};
+        llama_token_data_array single_token_data_array = { &single_token_data, 1, false };
+
+        // Apply grammar constraints to the single token
+        llama_sample_grammar(ctx_main, &single_token_data_array, ctx_sampling->grammar);
+
+        // Check if the token is valid according to the grammar by seeing if its logit has been set to -INFINITY
+        bool is_valid = single_token_data_array.data[0].logit != -INFINITY;
```
Comment (ejones) [inline common/sampling.cpp:235]: Yeah we could probably expose something more straightforward to check this if we want. It would probably be like `llama_grammar_accept_token` but returning a bool instead of throwing.

**Comment by ejones** [comment]
This is great. I toyed with an "optimistic" approach like this in the initial implementation, I think it's the way to go. My sense is that often the model will learn the grammar and tend to make good token decisions on its own, as you say.

> caching allowed vs unallowed tokens for each part of the grammar

Yeah I have the feeling that's next as well.

**Comment by kalomaze** [comment]
I was concerned that the code was maybe not working as expected but I believe I was wrong.

**Comment by kalomaze** [comment]
> I think it's the way to go

I think constraining the original set doesn't really provide much of a benefit, and in fact, it just makes the only candidate that is technically viable in some instances (specifically, the EOS token) more likely than it actually is, unless I'm wrong.

File: common/sampling.h [hunk 3]
```diff
@@ -92,16 +92,18 @@ std::string llama_sampling_print(const llama_sampling_params & params);
...
 llama_token llama_sampling_sample(
-        struct llama_sampling_context * ctx_sampling,
-        struct llama_context * ctx_main,
-        struct llama_context * ctx_cfg,
-        int idx = 0);
+    struct llama_sampling_context * ctx_sampling,
+    struct llama_context * ctx_main,
+    struct llama_context * ctx_cfg,
+    const int idx,
+    bool is_resampling = false);
```
[the remark this answers is not in the document]
Comment (kalomaze) [reply common/sampling.h:106]: Will look into this soon

**Comment by ggerganov** [comment]
I have missed that this PR has been updated.

@kalomaze Shall we merge this after fixing the conflict?

**Review by ggerganov — APPROVED** [review]

**Comment by AsimDerth** [comment]
And now grammar processing is slow again

**Comment by txbm** [comment]
I am also seeing significantly degraded performance when using a grammar for an inference that was consistently averaging about 7 minutes and now appears to be running more or less indefinitely. Can provide more details if helpful.
